grpc/src/js_client.rs

```diff
@@ -0,0 +1,370 @@
+    pub async fn get_balance(&self, address: String, denom: String) -> Result<JsCoin> {
+        self.client
+            .get_balance(&address.parse()?, denom)
+            .await
+            .map(Into::into)
+    }
+
+    /// Get balance of all coins
+    #[wasm_bindgen(js_name = getAllBalances)]
+    pub async fn get_all_balances(&self, address: String) -> Result<Vec<JsCoin>> {
+        self.client
+            .get_all_balances(&address.parse()?)
+            .await
+            .map(|coins| coins.into_iter().map(Into::into).collect())
+    }
+
+    /// Get balance of all spendable coins
+    #[wasm_bindgen(js_name = getSpendableBalances)]
+    pub async fn get_spendable_balances(&self, address: String) -> Result<Vec<JsCoin>> {
+        self.client
+            .get_spendable_balances(&address.parse()?)
+            .await
+            .map(|coins| coins.into_iter().map(Into::into).collect())
+    }
+
+    /// Get total supply
+    #[wasm_bindgen(js_name = getTotalSupply)]
+    pub async fn get_total_supply(&self) -> Result<Vec<JsCoin>> {
+        self.client
+            .get_total_supply()
+            .await
+            .map(|coins| coins.into_iter().map(Into::into).collect())
+    }
+
+    // TODO:
+    //  - cosmos.base.node
+    //  - cosmos.base.tendermint
+    //  - cosmos.tx
+    //  - celestia.blob
+    //  - celestia.core.tx
+}
+
+/// A helper to encode the SignDoc with protobuf to get bytes to sign directly.
+#[wasm_bindgen(js_name = protoEncodeSignDoc)]
+pub fn proto_encode_sign_doc(sign_doc: JsSignDoc) -> Vec<u8> {
+    SignDoc::from(sign_doc).encode_to_vec()
+}
+
+/// Signer that uses a javascript function for signing.
+pub struct JsSigner {
+    signer_fn: JsSignerFn,
+}
+
+impl DocSigner for JsSigner {
+    async fn try_sign(&self, doc: SignDoc) -> Result<Signature, SignatureError> {
+        let msg = JsSignDoc::from(doc);
+
+        let mut res = self.signer_fn.call1(&JsValue::null(), &msg).map_err(|e| {
+            let err = format!("Error calling signer fn: {e:?}");
+            SignatureError::from_source(err)
+        })?;
+
+        // if signer_fn is async, await it
+        if res.has_type::<Promise>() {
+            let promise = res.unchecked_into::<Promise>();
+            res = JsFuture::from(promise).await.map_err(|e| {
+                let err = format!("Error awaiting signer promise: {e:?}");
+                SignatureError::from_source(err)
+            })?
+        }
+
+        let sig = res.dyn_into::<Uint8Array>().map_err(|orig| {
+            let err = format!(
+                "Signature must be Uint8Array, found: {}",
+                orig.js_typeof().as_string().expect("typeof returns string")
+            );
+            SignatureError::from_source(err)
+        })?;
+
+        Signature::from_slice(&sig.to_vec()).map_err(SignatureError::from_source)
+    }
+}
+
+#[wasm_bindgen(typescript_custom_section)]
+const _: &str = "
+/**
+ * A payload to be signed
+ */
+export interface SignDoc {
```

Comment:
Since this name is coming from protobuf we probably shouldn't change it, but I find contracting "Document to be signed" into SignDoc confusing. :shrug: